refactor decorations to use selections

ianstormtaylor opened the pull request:
This changes how decorations are applied in Slate.

Previously, you would use a `decorate(text, block)` function in the schema. It would be called for each `text` node in the block, and you would add the marks to the characters yourself. But this led to issues because there are cases (like multi-line comments in syntax highlighting) where the decoration applies across multiple text nodes, and this system wasn't really built to handle that well.

Now, you use a `decorate(block)` function in the schema. It is no longer called for each of the text nodes. Instead of applying the new marks to the characters yourself, you return an array of selection objects, with associated marks. For example:

```js
{
  match: (object) => {
    return object.type == 'code_block',
  },
  decorate: (block) => {
    const text = block.getFirstText()

    return [{
      anchorKey: text.key,
      anchorOffset: 0,
      focusKey: text.key,
      focusOffset: 1,
      marks: [{ type: 'bold' }]
    }]
  }
}
```

This would turn add a `bold` mark to the first character of the matched blocks.

This would also make implementing things like multiple-cursor rendering easier, since the "decorations" now map exactly to the existing selections/ranges logic.

cc @Soreine, how does this sound? I think it should solve https://github.com/ianstormtaylor/slate/issues/893 and https://github.com/GitbookIO/slate-prism/pull/5.

lxcid commented:
This is awesome!

I do have question about does it improve dynamic decorator? When force rerender based on state not directly related to the editor (e.g. search result highlighting)

ianstormtaylor commented:
Hey @lxcid good question!

This PR itself doesn't support that, because the `decorate` logic in the schema is still based on node in the document themselves. 

But! It does get us closer to being able to do things like that. I was thinking of adding a `state.decorations` property and a `change.setDecorations(decorations)` method that would allow you do to that. That way you could change the decorations in the editor as the outside information changes (eg. someone performs a search) and then things will re-render.

I haven't tested the performance of doing that for large editors, since you'll be re-decorating the entire document at once, but I'm sure there are things we can do to improve it.

Soreine commented on packages/slate/src/models/node.js:
```diff
@@ -165,19 +165,12 @@ class Node {
     second = normalizeKey(second)
 
-    let sorted
+    const keys = this.getKeysAsArray()
```
Oh, here is the optimization :D

Soreine commented on packages/slate/src/models/schema.js:
```diff
@@ -126,24 +127,33 @@ class Schema extends Record(DEFAULTS) {
+      if (!rule.decorate) return
+      if (!rule.match(object)) return
+
+      const decorations = rule.decorate(object)
```
So the `decorate` calls themselves now live under a memoized function, which is nice :)

ianstormtaylor commented on packages/slate-react/src/components/text.js:
```diff
@@ -95,10 +90,19 @@ class Text extends React.Component {
+    const decs = decorations.filter((d) => {
+      const { startKey, endKey } = d
+      if (startKey == key || endKey == key) return true
+      const startsBefore = document.areDescendantsSorted(startKey, key)
```
I agree though, this can be improved.

One improvement I think would be a method that checks `(before, key, after)` or something about checking whether a node "overlaps" a range. Because here we have to do two separate checks which each iterate keys, etc. when really it could be handled in one go.

ianstormtaylor commented on packages/slate-react/src/plugins/core.js:
```diff
@@ -82,8 +81,8 @@ function Plugin(options = {}) {
-    const a = getPoint(native.anchorNode, native.anchorOffset, state, editor)
-    const f = getPoint(native.focusNode, native.focusOffset, state, editor)
+    const a = findPoint(native.anchorNode, native.anchorOffset, state)
+    const f = findPoint(native.focusNode, native.focusOffset, state)
```
Actually any place where we can doing `findPoint` for both anchor and focus can be slightly improved by checking if `collapsed` and if so only doing it once. But I'm not sure this is a bottleneck, so leaving for later.

ianstormtaylor commented on packages/slate/src/models/node.js:
```diff
@@ -958,18 +925,29 @@ class Node {
    */
 
-  getKeys() {
+  getKeysAsArray() {
```
I think this method can also be improved in performance if it was recursive, such that each nested child's keys would be cached, instead of having to iterate the entire tree each time the document changes.

ianstormtaylor commented:
Actually @lxcid I've just implemented that behavior that you're asking about! I added `state.decorations`, which can be set based on any outside information. Along with a simple `search-highlighting` example to show you how it works.

lxcid commented:
Wow! Thank for considering my request @ianstormtaylor and getting it in so soon! I can't wait to use it soon! The demo looks great!
